router/internal/eventorder/eventorder.go

```diff
@@ -0,0 +1,248 @@
+package eventorder
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/rudderlabs/rudder-server/jobsdb"
+)
+
+var ErrUnsupportedState = errors.New("unsupported state")
+
+// NewBarrier creates a new properly initialized Barrier
+func NewBarrier(abortConcurrencyLimit int) *Barrier {
+	return &Barrier{
+		concurrencyLimit: abortConcurrencyLimit,
+		barriers:         make(map[string]*barrierInfo),
+	}
+}
+
+type Barrier struct {
+	concurrencyLimit int
+	mu               sync.RWMutex
+	queue            []command
+	barriers         map[string]*barrierInfo
+}
+
+// Enter the barrier for this userID and jobID. If there is not already a barrier for this userID
+// returns true, otherwise false along with the previous failed jobID if this is the cause of the barrier.
+// Another scenario where a barrier might exist for a user is when the previous job has failed in an unrecoverable manner and the concurrency limiter is enabled.
+func (b *Barrier) Enter(userID string, jobID int64) (accepted bool, previousFailedJobID *int64) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	barrier, ok := b.barriers[userID]
+	if !ok {
+		return true, nil // no barrier, accept
+	}
+
+	// if there is a failed job in the barrier, only this job can enter the barrier
+	if barrier.failedJobID != nil {
+		failedJob := *barrier.failedJobID
+		return jobID == failedJob, &failedJob
+	}
+
+	// if the concurrency limiter is enabled, only the configured number of concurrent jobs can enter the barrier
+	if barrier.concurrencyLimiter != nil {
```

Comment:
[optional] early return
```suggestion
	if barrier.concurrencyLimiter == nil {
		return true, nil
	}
```